doc/internals.rst

```diff
@@ -231,3 +231,448 @@ re-open it directly with Zarr:
+
+.. code-block::
+
+    setuptools.setup(
+        entry_points={
+            "xarray.backends": [
+                "engine_name=your_package.your_module:your_backendentrypoint"
+            ],
+        },
+    )
+
+in ``setup.cfg``:
+
+.. code-block:: cfg
+
+    [options.entry_points]
+    xarray.backends =
+        engine_name = your_package.your_module:your_backendentrypoint
+
+
+See https://packaging.python.org/specifications/entry-points/#data-model
+for more information
+
+.. _RST lazy_loading:
+
+How to support Lazy Loading
++++++++++++++++++++++++++++
+If you want to make your backend effective with big datasets, then you should
+support lazy loading.
+Basically, you shall replace the :py:class:`numpy.ndarray` inside the
+variables with a custom class that supports lazy loading indexing.
+See the example below:
+
+.. code-block:: python
+
+    backend_array = MyBackendArray()
+    data = indexing.LazilyOuterIndexedArray(backend_array)
+    var = xr.Variable(dims, data, attrs=attrs, encoding=encoding)
+
+Where:
+
+- :py:class:`~xarray.core.indexing.LazilyOuterIndexedArray` is a class
+  provided by Xarray that manages the lazy loading.
+- ``MyBackendArray`` shall be implemented by the backend and shall inherit
+  from :py:class:`~xarray.backends.common.BackendArray`.
+
+BackendArray subclassing
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+The BackendArray subclass shall implement the following method and attributes:
+
+- the ``__getitem__`` method that takes in input an index and returns a
+  `NumPy <https://numpy.org/>`__ array
+- the ``shape`` attribute
+- the ``dtype`` attribute.
+
+
+Xarray supports different type of
+`indexing <http://xarray.pydata.org/en/stable/indexing.html>`__, that can be
+grouped in three types of indexes
+:py:class:`~xarray.core.indexing.BasicIndexer`,
+:py:class:`~xarray.core.indexing.OuterIndexer` and
+:py:class:`~xarray.core.indexing.VectorizedIndexer`.
+This implies that the implementation of the method ``__getitem__`` can be tricky.
+In oder to simplify this task, Xarray provides a helper function,
+:py:func:`~xarray.core.indexing.explicit_indexing_adapter`, that transforms
+all the input  ``indexer`` types (`basic`, `outer`, `vectorized`) in a tuple
+which is interpreted correctly by your backend.
+
+This is an example ``BackendArray`` subclass implementation:
+
+.. code-block:: python
+
+    class MyBackendArray(BackendArray):
+        def __init__(
+            self,
+            shape,
+            dtype,
+            lock,
+            # other backend specific keyword arguments
+        ):
+            self.shape = shape
+            self.dtype = lock
+            self.lock = dtype
+
+        def __getitem__(self, key):
+            return indexing.explicit_indexing_adapter(
+                key,
+                self.shape,
+                indexing.IndexingSupport.BASIC,
+                self._raw_indexing_method,
+            )
+
+        def _raw_indexing_method(self, key):
```

Comment:
```suggestion
        def _raw_indexing_method(self, key: tuple) -> np.typing.ArrayLike:
```

(could also note that in this case, `key` could be `Tuple[Union[int, slice], ...]` because it only implements "basic" indexing)